tests: Timer behavior test for TICKS_PER_SEC being sensible

teburd opened the pull request:
Test timers with a train of one tick timers to test that
a configured SYS_CLOCK_TICKS_PER_SEC is sensible. If the TICKS_PER_SEC
is too high the timer train will take longer than expected to reach
the station. Or worse, if the timer driver has too short of a minimum
delay for its processing power and the tick rate is too high its
possible the device will get caught in an interrupt loop causing threads
to not run at all while processing timers scheduled too close together.


Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
While I describe it as a timer train (and it can be with TIMERS configured to more than 1) its setup as a single timer scheduled absolutely every tick with the expectation that this can be done on time with other work occurring in the test thread.

andyross commented on tests/kernel/timer/timer_behavior/src/tick_timer_train.c:
```diff
@@ -0,0 +1,143 @@
+		}
+		if (remaining_timers == 0) {
+			break;
+		}
```
This loop took me a while to figure out.  How about this, which is simpler and easier to understand:

```c
while (remaining_timers > 0) {
    if (k_sem_take(&timers_sem, K_NO_WAIT) == 0) {
        remaining_timers--;
    }
    busy_loops++;
    k_busy_wait(250000);
}

andyross replied:
Sorry, didn't update the comment for the next one I wrote: you'd want the busy_wait() in this version to be much faster (or just absent, it should be OK to spin on a semaphore).

andyross replied:
Or another option would be to spawn one low priority thread that just does a "`while(true);`" and just do this with for loop over the known number of timer events.

teburd replied:
I want to validate that the thread is in fact running and that it isn't really prevented from running for any significant duration, the time period isn't a big deal. See how busy_loops is used below to validate that yes, the thread ran, and it ran for a reasonable amount of time despite the timers being maxed out.

teburd replied:
The way the test is setup, I'm testing that timers and its machinery don't take more than give or take 10% of the time available even if scheduled on every tick. Perhaps that's too generous to threads, but that's probably reasonable given most work is really meant to happen in threads and spending more time in timer machinery than threads is a net negative in the worst case scenario setup here. Hopefully that makes it clearer, I'm not sure the thread spawned up with while(true) really shows that where as the k_busy_wait and count do.

andyross commented on tests/kernel/timer/timer_behavior/src/tick_timer_train.c:
```diff
@@ -0,0 +1,143 @@
+		TC_PRINT("Faking busy work, remaining timers is %u, timer callbacks %u\n",
+			 remaining_timers, timers[0].callbacks);
+		busy_loops++;
+		k_busy_wait(250000);
```
This time looks wrong.  250ms is much, much longer than a tick, why wait so long when the semaphore gives should be arriving 10-100x faster?

teburd replied:
> Some notes. On the whole I think the test is a great idea, but I do caution that this is all but guaranteed to produce a flood of false positives on SMP qemu platforms in CI. Timer jitter due to host scheduling is a known (and largely unresolvable) thing, and this is a flashlight shining right on the problem.

Well the test case does try and exclude simulators already in testcase.yaml by tying it to the the existing timer behavior test. Though that has other issues. On arm the order of the tests (I need to see if I can change the ordering...) seems to be this timer train test followed by the behavior test, but by then the timer counter starts rolling over and the jitter_drift test fails because of that.

andyross commented on tests/kernel/timer/timer_behavior/src/main.c:
```diff
@@ -4,197 +4,6 @@
-
-#include <zephyr/kernel.h>
-
-#include <zephyr/tc_util.h>
```
Can you split the code motion out of main and new code with the new test into two different patches for clarity?

andyross commented on tests/kernel/timer/timer_behavior/src/tick_timer_train.c:
```diff
@@ -0,0 +1,143 @@
+		int64_t next = tm_wrap->last_scheduled + TIMERS;
+
+		tm_wrap->last_scheduled = next;
+		k_timer_start(tm, K_TIMEOUT_ABS_TICKS(next), K_NO_WAIT);
```
Given that the provenance of the test is to exercise timer agility, I don't know that you want to be doing this much work in the callback.  Why not just use a periodic timer with a 1 tick period?  Basically: there are faster ways to do this.  Since this is a performance test, we should be exercising the fast paths that performance-sensitive apps will take.

teburd replied:
I tried doing that, but it always took twice as long as I expected, I think because of the + 1 in timer.c for periodically scheduled timers was my thought but I didn't look that hard at it to be honest.

andyross reviewed (commented):
Some notes.  On the whole I think the test is a great idea, but I do caution that this is all but guaranteed to produce a flood of false positives on SMP qemu platforms in CI.  Timer jitter due to host scheduling is a known (and largely unresolvable) thing, and this is a flashlight shining right on the problem.

Commit:
tests: Rename timer behavior main.c

Move the main.c timer behavior test code to jitter_drift.c
so that other tests may be added to the suite.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
tests: Timer behavior custom test_main

Adds a custom test_main and renames the test suite for jitter_drift.

Runs the jitter_drift test suite.

The order of these tests matter on hardware as the counter is often
reset on loading the test program. This is useful as its far less likely
to encounter a clock counter rollover. On arm this is especially useful.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
tests: Tick rate testing with timer train

Test timers with a train of one tick timers to test that
a configured SYS_CLOCK_TICKS_PER_SEC is sensible. If the TICKS_PER_SEC
is too high the timer train will take longer than expected to reach
the station. Worse, if the timer driver has too short of a minimum
delay for its processing power and the tick rate is too high its
possible the device will get caught in an interrupt loop
preventing any threads from running while processing timers.

This test validates that the tick rate configured is actually able to be
processed without delays while also having work done in threads ensuring
that no thread scheduling delays occur either from delayed timers or an
interrupt loop from preventing threads from running.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

peter-mitsis commented on tests/kernel/timer/timer_behavior/src/tick_timer_train.c:
```diff
@@ -0,0 +1,145 @@
+
+	TC_PRINT("Starting Timers with Skews\n");
+	int64_t tick = k_uptime_ticks();
+
+	for (int i = 0; i < TIMERS; i++) {
+		timers[i].last_scheduled = tick+1;
+		k_timer_start(&timers[i].tm, K_TIMEOUT_ABS_TICKS(tick + 1), K_NO_WAIT);
+	}
```
Question: Is it worth aligning to a tick boundary prior to starting the timer(s)?

andyross replied:
Or just skip the analysis for the first pair of times; the timer hooks will self-align by definition anyway.

teburd replied:
Probably not a big deal, they should self align and the test is try to ensure a mean usefulness with maximal timer scheduling. This won’t change that I don’t think.

I do print out the maximum cycle delta but that’s really not that useful probably, one tempted to remove it.

teburd commented:
Ok I sorted out the issues, so I can test with N timers in the timer train now and things pass just fine on platforms I'd expect to pass (cortex-m7 on tdk_robokit1, cavs25) with 4 timers.

teburd commented:
@andyross @peter-mitsis any other thoughts here? I think some of the review comments above are maybe a mis-interpretation of the intent, especially with regards to the k_busy_wait, but hopefully those are better explained in the comments now. If not can we discuss what it is needs to change to make this a test in the tree?

peter-mitsis reviewed (dismissed):
No objections from me.

andyross reviewed (dismissed):
Looks good.  Still a little worried about the apparent need for that logic in the timer callback vs. just using a periodic timer.  Might be hiding subsystem or platform bugs.

carlescufi commented:
@teburd please rebase

andyross reviewed (approved):
refresh +1

(I continue to hate this, can't we go back to the "keep reviews" setting?  Or, I dunno, implement an honor system "all I did was trivial fixups/rebases/spellchecking" button submitters can use?)
